pilot/pkg/config/kube/gateway/conditions.go

```diff
@@ -0,0 +1,206 @@
+			// TODO this is not namespaced but a namespace is required
+			ref.Namespace = "default"
+		} else {
+			s := strings.Split(gw, "/")
+			ref.Name = s[1] // TODO name here is the internal name, need to use the external name
+			ref.Namespace = s[0]
+		}
+		var condition metav1.Condition
+		if routeErr != nil {
+			condition = metav1.Condition{
+				Type:               string(k8s.ConditionRouteAdmitted),
+				Status:             kstatus.StatusFalse,
+				ObservedGeneration: obj.Generation,
+				LastTransitionTime: metav1.Now(),
+				Reason:             routeErr.Reason,
+				Message:            routeErr.Message,
+			}
+		} else {
+			condition = metav1.Condition{
+				Type:               string(k8s.ConditionRouteAdmitted),
+				Status:             kstatus.StatusTrue,
+				ObservedGeneration: obj.Generation,
+				LastTransitionTime: metav1.Now(),
+				Reason:             "RouteAdmitted",
+				Message:            "Route was valid",
+			}
+		}
+		gws = append(gws, k8s.RouteGatewayStatus{
+			GatewayRef: ref,
+			Conditions: []metav1.Condition{condition},
+		})
+	}
+	return gws
+}
+
+type ConfigErrorReason = string
+
+const (
+	// InvalidDestination indicates an issue with the destination
+	InvalidDestination ConfigErrorReason = "InvalidDestination"
+	// InvalidFilter indicates an issue with the filters
+	InvalidFilter ConfigErrorReason = "InvalidFilter"
+	// InvalidTLS indicates an issue with TLS settings
+	InvalidTLS ConfigErrorReason = "InvalidTLS"
+	// InvalidConfiguration indicates a generic error for all other invalid configurations
+	InvalidConfiguration ConfigErrorReason = "InvalidConfiguration"
+)
+
+// ConfigError represents an invalid configuration that will be reported back to the user.
+type ConfigError struct {
+	Reason  ConfigErrorReason
+	Message string
+}
+
+type condition struct {
+	// reason defines the reason to report on success. Ignored if error is set
+	reason string
+	// message defines the message to report on success. Ignored if error is set
+	message string
+	// status defines the status to report on success. The inverse will be set if error is set
+	// If not set, will default to StatusTrue
+	status metav1.ConditionStatus
+	// error defines an error state; the reason and message will be replaced with that of the error and
+	// the status inverted
+	error *ConfigError
+}
+
+// setConditions sets the existingConditions with the new conditions
+func setConditions(generation int64, existingConditions []metav1.Condition, conditions map[string]*condition) []metav1.Condition {
+	// Sort keys for deterministic ordering
+	condKeys := make([]string, 0, len(conditions))
+	for k := range conditions {
+		condKeys = append(condKeys, k)
+	}
+	sort.Strings(condKeys)
+	for _, k := range condKeys {
+		cond := conditions[k]
+		if cond.error != nil {
+			existingConditions = kstatus.ConditionallyUpdateCondition(existingConditions, metav1.Condition{
+				Type:               k,
+				Status:             kstatus.InvertStatus(cond.status),
```

Comment:
Why do we do this? In the other branch it seems like we prefer "True". I'm guessing here we just prefer false? What if the condition starts as false, is it valid to make it "True"?

Reply:
Ah. So in the other branch it's invalid if the condition isn't populated, but it _should_ be "false" for no errors. 

I still don't understand why we just invert here with no check. Don't we want `Status = false` when there is an error?

Reply:
The problem is a condition can be "negative polarity" (ex: ListenerInvalid) or "positive polarity" (ex: ListenerValid). Lots of info about this in https://github.com/kubernetes/community/blob/4c9ef2d/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties.

So there is no global default that we want. Instead, I had each `condition` define their default positive status, and if there is an error we invert it